[ASCII-1549] improve the overview and creating a component page (#25028)
DataDog/datadog-agent: docs/public/components/overview.md

## What is a Component?

> A component encapsulates a particular piece of logic/feature in a clear and documented interface. 

Reviewer: I'd like to highlight the encapsulate part:

```suggestion
The goal of a component is to encapsulates a particular piece of logic/feature and provide a clear and documented interface. 
```

?

A component must:

  + The component hides the complexity of the implementation from its users.
  + All components must be reusable no matter the context or the binary in which they're included.
  + A component must be tested.
  + Expose a mock implementation to help testing.
  + Any change within a component, that don't change its interface, should not require QA of the other component using
    it.

>   + Each component should be owned by a single team which will support and maintain it.

Author: Providing a more condense list is best to capture the reader attention. In the creating a component page we can mention any other point that is missing from the overview page 😄

Reviewer: I felt that the 3 sections highlighted clearly why we're doing the whole component thing. Explaining what must be done it good, but we're trying to convince people, so the why is also important.

Since each component is an interface to the outside we can have several implementations for it.

## Grouping components: Bundles

There is a large and growing number of components, and listing those components out repeatedly could grow tiresome and
cause bugs. Component bundles provide a way to manipulate multiple components, usually at the meta or service level, as
a single unit. For example, while Logs-Agent is internally composed of many components, those can be used as a single unit
with a `logs` bundle.

As an aid to thinking about this question, consider the following levels:

1. Large-scale parts of the Agent that agglomerate multiple components would be a `Bundle`. Example: the Logs-Agent.
2. Dedicated section or logic block that can be reused in multiple places would be a `Component`. Example: the logger
   component.
3. A specific version of a component would be an implementation. Example: a component in charge of compressing payload
   could have a ZSTD and ZIP implementation. Most components will only have a single implementation.

## Stateless logic

Any logic/helper that is stateless (no globals and not `init` function) can but doesn't need to be a component. Such

> basic helpers can remain or be added to the `pkg` folder.

Author: I think mentioning bundles in the overview might be a bit overkill as is not a core concept of a component

Reviewer: Let's move the `Grouping components: Bundles` to the bundle section then.

> basic helpers can remain or be added to the `pkg` folder.

Reviewer: Why not keep this ?

Author: We should add this section in the FAQ

> Components implementation can be exported as a Go module. We will see in more detail how to create components in the next section.

Reviewer 2: ```suggestion
Component implementations can be exported as Go modules. The next section explains in more detail how to create components.
```

> methods, and increased reusability by separating logical units into components. **`Fx` itself is not intrinsic to the

Reviewer: I think a quick explanation of what FX is used for was useful in this context, no ?

Author: I agree. Let's have a disclaimer about components not being 1 to 1 to Fx
